lms/djangoapps/course_blocks/transformers/user_partitions.py

```diff
@@ -0,0 +1,226 @@
+            lambda block_key: not block_structure.get_transformer_block_field(
+                block_key, self, 'merged_group_access'
+            ).check_group_access(user_groups)
+        )
+
+
+class MergedGroupAccess(object):
+    """
+    A class object to represent the computed access value for a block,
+    merged from the inherited values from its ancestors.
+
+    How group access restrictions are represented within an XBlock:
+      - group_access not defined                           => No group access restrictions.
+      - For each partition:
+          - partition.id not in group_access               => All groups have access for this partition
+          - group_access[partition_id] is None             => All groups have access for this partition
+          - group_access[partition_id] == []               => All groups have access for this partition
+          - group_access[partition_id] == [group1..groupN] => groups 1..N have access for this partition
+
+    We internally represent the restrictions in a simplified way:
+      - self._access == {}                                 => No group access restrictions.
+      - For each partition:
+          - partition.id not in _access                    => All groups have access for this partition
+          - _access[partition_id] == set()                 => No groups have access for this partition
+          - _access[partition_id] == set(group1..groupN)   => groups 1..N have access for this partition
+
+    Note that a user must have access to all partitions in group_access
+    or _access in order to access a block.
+    """
+    def __init__(self, user_partitions, xblock, merged_parent_access_list):
+        """
+        Arguments:
+            user_partitions (list[UserPartition])
+            xblock (XBlock)
+            merged_parent_access_list (list[MergedGroupAccess])
+        """
+
+        self._access = {}  # { partition.id: set(IDs of groups that can access partition }
+
+        # Get the group_access value that is directly set on the xblock.
+        # Do not get the inherited value since field inheritance doesn't
+        # take a union of them for DAGs.
+        block_group_access = get_field_on_block(xblock, 'group_access', default_value={})
+
+        for partition in user_partitions:
+
+            # Within this loop, None <=> Universe set <=> "No access restriction"
+
+            block_group_ids = set(block_group_access.get(partition.id, [])) or None
+            parents_group_ids = []
+            for merged_parent_access in merged_parent_access_list:
+                # pylint: disable=protected-access
+                if partition.id in merged_parent_access._access:
+                    parents_group_ids.append(merged_parent_access._access[partition.id])
+                else:
+                    parents_group_ids = []
+                    break
+
+            merged_parent_group_ids = (
+                set().union(*parents_group_ids)
```

Comment:
Good question.  We want to collapse each of the lists from each of the parents.  So, `parents_group_ids` is actually a list of lists.  I'll add a comment to clarify this.